host-controller/src/main/java/org/jboss/as/host/controller/operations/StartServersHandler.java

```diff
@@ -95,27 +96,47 @@ public void execute(OperationContext context, ModelNode operation) throws Operat
                 final Resource resource =  context.readResource(PathAddress.EMPTY_ADDRESS);
                 final ModelNode hostModel = Resource.Tools.readModel(resource);
                 if(hostModel.hasDefined(SERVER_CONFIG)) {
+
                     final ModelNode servers = hostModel.get(SERVER_CONFIG).clone();
+                    final boolean restartRequired = runningModeControl.getRestartMode() == RestartMode.SERVERS;
+                    final boolean blockUntilStopped = hostControllerEnvironment.isRestart() || !runningModeControl.isReloadByHandler() || (runningModeControl.isReloadByHandler() && runningModeControl.isBlockUntilStopped());
+
+                    if (blockUntilStopped){
+                        reconnectAndWaitStoppingServers(servers, domainModel, true);
+                    }
+
                     if (hostControllerEnvironment.isRestart() || runningModeControl.getRestartMode() == RestartMode.HC_ONLY){
-                        restartedHcStartOrReconnectServers(servers, domainModel, context);
-                        runningModeControl.setRestartMode(RestartMode.SERVERS);
+                        restartedHcStartOrReconnectServers(servers, domainModel, context, restartRequired);
                     } else {
-                        cleanStartServers(servers, domainModel, context);
+                        cleanStartServers(servers, domainModel, context, restartRequired);
                     }
                 }
+                //Reset runningModeControl values to default for the values that cannot be reset at ProcessReloadHandler.ReloadContext.doReloado
+                runningModeControl.setRestartMode(RestartMode.SERVERS);
+                runningModeControl.setBlockUntilStopped(true);
                 context.completeStep(OperationContext.RollbackHandler.NOOP_ROLLBACK_HANDLER);
             }
         }, OperationContext.Stage.RUNTIME);
     }
 
-    private void cleanStartServers(final ModelNode servers, final ModelNode domainModel, OperationContext context) throws OperationFailedException {
+    private void reconnectAndWaitStoppingServers(final ModelNode servers, final ModelNode domainModel, boolean blockUntilStopped){
```

Comment:
I believe the 'blockUntilStopped' param is always 'true' and therefore doesn't need to be a param.

Reply:
Now `blockUntilStopped` is always `runningModeControl.isBlockUntilStopped()`, by default always true, false if the user wants, so it makes more sense now,  @bstansberry do you agree?